runners/direct-java/src/main/java/org/apache/beam/runners/direct/WriteWithShardingFactory.java

```diff
@@ -0,0 +1,132 @@
+import org.apache.beam.sdk.transforms.windowing.DefaultTrigger;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import org.joda.time.Duration;
+
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * A {@link PTransformOverrideFactory} that overrides {@link Write} {@link PTransform PTransforms}
+ * with an unspecified number of shards with a write with a specified number of shards. The number
+ * of shards is uniformly randomly distributed between 3 and 7 shards.
+ */
+class WriteWithShardingFactory implements PTransformOverrideFactory {
+  @Override
+  public <InputT extends PInput, OutputT extends POutput> PTransform<InputT, OutputT> override(
+      PTransform<InputT, OutputT> transform) {
+    if (transform instanceof Write.Bound) {
+      Write.Bound<InputT> that = (Write.Bound<InputT>) transform;
+      if (that.getNumShards() == 0) {
+        return (PTransform<InputT, OutputT>) new DynamicallyReshardedWrite<InputT>(that);
+      }
+    }
+    return transform;
+  }
+
+  private static class DynamicallyReshardedWrite <T> extends PTransform<PCollection<T>, PDone> {
+    private final transient Write.Bound<T> original;
+
+    private DynamicallyReshardedWrite(Bound<T> original) {
+      this.original = original;
+    }
+
+    @Override
+    public PDone apply(PCollection<T> input) {
+      PCollection<T> records = input.apply("RewindowInputs",
+          Window.<T>into(new GlobalWindows()).triggering(DefaultTrigger.of())
+              .withAllowedLateness(Duration.ZERO)
+              .discardingFiredPanes());
+      final PCollectionView<Long> numRecords = records.apply(Count.<T>globally().asSingletonView());
+      PCollection<T> resharded =
+          records
+              .apply(
+                  "ApplySharding",
+                  ParDo.withSideInputs(numRecords)
+                      .of(new KeyBasedOnCountFn<T>(numRecords)))
+              .apply("GroupIntoShards", GroupByKey.<Long, T>create())
+              .apply("DropShardingKeys", Values.<Iterable<T>>create())
+              .apply("FlattenShardIterables", Flatten.<T>iterables());
+      // This is an inverted application to apply the expansion of the original Write PTransform
+      // without adding a new Write Transform Node, which would be overwritten the same way, leading
+      // to an infinite recursion. We cannot modify the number of shards, because that is determined
+      // at runtime.
+      return original.apply(resharded);
+    }
+  }
+
+  @VisibleForTesting
+  static class KeyBasedOnCountFn<T> extends DoFn<T, KV<Long, T>> {
```

Comment:
s/Long/Integer/. If we get a job where we get to the write phase having generated 10^2.1B elements in the direct runner, I'll eat my Beam hat.